### PR TITLE
Submit hackathon result for review.

### PR DESCRIPTION
**Pull request by jstoone**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    ],
+    "require": {
+        "php": ">=5.5.9"
+    },
+    "require-dev": {
+        "phpunit/phpunit": "~4.8"
+    },
+    "autoload": {
+        "psr-0": {
+            "JakobSteinn\\HTMLFormValidator": "src/"
```
Comment (samsonasik) [inline composer.json:19]: add \ after HTMLFormValidator

Comment (samsonasik) [inline composer.json:19]: I mean \\

File: .gitignore [hunk 2]
```diff
@@ -0,0 +1,13 @@
+/vendor
+
+*.swp
+.idea
```
Comment (Ocramius) [inline .gitignore:4]: These should be in your [global gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore)

File: README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+            name="email"
+            value=""
+            data-validator="email"
+            required="required"
+        />
+        <input type="submit"/>
+    </form>
+';
+
+$formValidator = (new Validator($formHTML))->validate($_POST);
```
Comment (Ocramius) [inline README.md:23]: `$formValidation` - I was suggesting having a value object as return value here ;-)

Something like:

``` php
$validationResult = $validator->validate($_POST);

var_dump($validationResult->isSuccess());
var_dump($validationResult->getErrorMessages());
var_dump($validationResult->getRawInputValues());
var_dump($validationResult->getValidValues());
```

File: src/JakobSteinn/HTMLFormValidator/RuleProviders/AbstractRuleProvider.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
+<?php
+
+namespace JakobSteinn\HTMLFormValidator\RuleProviders;
+
+abstract class AbstractRuleProvider implements RuleProviderContract
```
Comment (Ocramius) [inline src/JakobSteinn/HTMLFormValidator/RuleProviders/AbstractRuleProvider.php:5]: This is actually exactly what `Zend\Validator` does (and the symfony validators, and respect validation, and others): I would suggest not re-designing that, since the validation lib is quite solid there.

Actually, you may just leverage [`Zend\InputFilter\Factory`](http://framework.zend.com/manual/current/en/modules/zend.input-filter.intro.html) to initialize an array of validators (and just use this lib to extract the array data to pass to that factory)

File: src/JakobSteinn/HTMLFormValidator/Validator.php [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+     * @var DOMDocument
+     */
+    protected $formDocument;
+
+    /**
+     * The rule providers used to validate input
+     * 
+     * @var JakobSteinn\HTMLFormValidator\RuleProviders\AbstractRuleProvider
+     */
+    protected $ruleProviders = [
```
Comment (Ocramius) [inline src/JakobSteinn/HTMLFormValidator/Validator.php:24]: This is indeed what the [`ValidatorPluginManager`](https://github.com/zendframework/zend-validator/blob/33f0141fce05155fb75b5d68fac5856d3a21ebd2/src/ValidatorPluginManager.php#L32) does in zf2 (sorry if I always use zf2 as an example, it's just that that is my area of expertise)

File: src/JakobSteinn/HTMLFormValidator/Validator.php [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+
+    /**
+     * Determine whether the validation has passed, returns true
+     * if no errors, else it returns appropriate errors.
+     * 
+     * @return bool|array
+     */
+    public function passes()
+    {
+        if (count($this->errors) > 0) {
```
Comment (Ocramius) [inline src/JakobSteinn/HTMLFormValidator/Validator.php:92]: This bit of logic makes your validator stateful and hard to re-use over multiple executions: if the validator pushes all the results to an immutable value object, then we get a better/cleaner API

File: src/JakobSteinn/HTMLFormValidator/Validator.php [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    /**
+     * Determine if the given data is given, since it is required
+     *
+     * @param  string 	$field     The field under validation
+     * @param  array 	$fieldData The data associated with the given field
+     * @return void
+     */
+    protected function validateRequiredFor($field, $fieldData = null)
```
Comment (Ocramius) [inline src/JakobSteinn/HTMLFormValidator/Validator.php:131]: use `private` whenever possible: you will avoid hairy situations with inheritance

File: src/JakobSteinn/HTMLFormValidator/Validator.php [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    /**
+     * Find the first form on page, and parse <input>-elements
+     *
+     * @return array
+     */
+    protected function parseForRules()
+    {
+        $form = $this->formDocument->getElementsByTagName('form')->item(0);
```
Comment (Ocramius) [inline src/JakobSteinn/HTMLFormValidator/Validator.php:147]: This is a good first version, but indeed a bit naive. This is where the library needs overhauling in future, and therefore I suggest splitting out this bit into a collaborator that can be made more flexible in future

**Comment by Ocramius** [comment]
Overall, the idea is great and it reflects my thoughts in the initial proposal. Thanks for following on that!

What has to be fixed (IMO) is that you really don't want to maintain validators yourself.

Validation is a complex and security-sensitive context, and therefore I'd leave it to symfony/zendframework (since their security teams rock), and I'd just reuse their validators in here, referencing them by name (and failing eagerly when a validator is not recognized).

Also, you really want to support any valid HTML form, with any nesting level, therefore some XPath logic is required here.

I suggest looking at how https://github.com/minkphp/Mink does it, and how behat in general attempts to interact with forms (for example in https://github.com/Behat/Mink/blob/d99b94af22d34b12d4417512538f13890c7fa5be/src/Behat/Mink/Element/TraversableElement.php#L145 )

**Comment by jstoone** [comment]
Thank you for the great code review @Ocramius! I am really astonished by the level of quality, and if this is the level of dedication then I should for sure do more OS projects based on proposals. :)
Also nice of you to point out example replacements for each piece (zf2/validator and mink), that reduced my research a lot. 

I'm glad that I am heading in the right direction, even though most of the code needs a total rewrite.

I'm closing this pull request and submitting a new one with a revised approach utilizing the review items you've submitted.

In short the new pull request will feature:
- Extracted validator, utilizing zf2:
  - (ValidatorPluginManager)[https://github.com/zendframework/zend-validator/blob/33f0141fce05155fb75b5d68fac5856d3a21ebd2/src/ValidatorPluginManager.php]
  - (Zend/InputFilter/Factory)[http://framework.zend.com/manual/current/en/modules/zend.input-filter.intro.html]
- Leverage a higher level package for traversing through the DOM, and finding a form based on selector
- Refactor form validator so that it returns a VO, that user can query for validation status

**Comment by yourwebmaker** [comment]
I suggest you to use some validation library to preforms validation instead building a new one.

Use this library only for catch a form string and then sends to the user validation messages.

**Comment by Ocramius** [comment]
@yourwebmaker that's already what he said he'll be doing ;-)
